File: 프로그래머스/2/42577. 전화번호 목록/전화번호 목록.py
def solution(phone_book):
    hash_map = {}
    
    for phone in phone_book:
        hash_map[phone] = 1
        
    for phone in phone_book:
        prefix = ""
        
        # 맨 마지막 원소 제외해서 보기
        # 자기자신을 체크해선 안 되고,
        # 같은 전번이 중복되지 않는다는 조건이 있기 때문
        # 역순은 [::-1]
        for num in phone[:-1]:
            prefix += num
            if prefix in hash_map:
                return False
            
    return True

File: 프로그래머스/2/42577. 전화번호 목록/전화번호 목록.js
function solution(phone_book) {
    const hash_map = {};
    
    for (const phone of phone_book) {
        hash_map[phone] = 1;
    }
    
    for (const phone of phone_book) {
        let prefix = "";
        for (let i = 0; i < phone.length - 1; i++) {
            prefix += phone[i];
            if (hash_map[prefix]) {
                return false;
            }
        }
    }
    
    return true;
}